### PR TITLE
full rewrite in script syntax

### PR DESCRIPTION
**Pull request by wpdebruin**
ResourceService: rewrite in script
i18n: rewrite in script, several multiUsers fixed, for loops replaced by array and structkey iterations

### CODE REVIEW

**Comment by wpdebruin** [comment]
STUPID parser bug in acf2016
When I try this
function getResource(
        required resource, 
        default, 
        locale = variables.i18n.getfwLocale(), 
        values, 
        bundle = "default"
    ) {
I get a parsing error on the default  . As soon as I change it to defaultA errors are gone :sob:. But that’s not the name of this parameter… So script syntax sucks here in 2016.

The only solution I see here is remove it from the function declaration.  Or revert to some include which has the function in script. YIKES! Other solutions are welcome. All builds are Ok now, but I don’t like it….
I read default is a reserved word in script syntax in ACF. So that’s probably why it fails.

**Comment by wpdebruin** [comment]
Ok, I did some more investigation on this default param stuff. And I think you have hit this limitation before, because you already renamed the param in the mixin to defaultValue . So getResource in the mixin has different param naming from the resourceService itself. My suggestion would be to just rename the param to defaultValue in the ResourceService. Probably nobody notices because there’s hardly a reason to use this param, and even less reason to use this param by name. And if I am going to apply all hierarchial Resources stuff, we can call it v2 and make it a (hardly) breaking change. So what do yo think?

**Comment by wpdebruin** [comment]
I finally found a decent fix for the `default` vs `defaultValue` argument issue.
default or defaultValue is not required, so I added defaultValue to the function definition.
In the code I check for defaultValue first and if it is not there I check the default argument, so it remains backwards compatible. In the mixin (which was using defaultValue anyway) the renaming of params has been removed, it is not necessary anymore.

**Comment by lmajano** [comment]
Will since this is a major release it’s ok to break compatibility.

Luis Majano
CEO
Ortus Solutions, Corp
www.ortussolutions.com
P/F: 1-888-557-8057
________________________________

File: models/ResourceService.cfc [hunk 1]
```diff
@@ -1,439 +1,517 @@
...
+		// store controller variable
+		variables.controller = arguments.controller;
+		variables.i18n       = arguments.i18n;
+
+		// check if localization struct exists in memory, else create it.
+		if ( !arguments.controller.settingExists( "RBundles" ) ) {
+			arguments.controller.setSetting( "RBundles", {} );
+		}
+		// store bundles
+		variables.aBundles              = arguments.controller.getSetting( "RBundles" );
```
Comment (lmajano) [inline models/ResourceService.cfc:39]: this is duplicated below

File: models/ResourceService.cfc [hunk 2]
```diff
@@ -1,439 +1,517 @@
...
 					// load a bundle and store it.
-					instance.aBundles[ arguments.rbAlias ][ arguments.rbLocale ] = getResourceBundle(rbFile=arguments.rbFile, rbLocale=arguments.rbLocale);
+					variables.aBundles[ arguments.rbAlias ][ arguments.rbLocale ] = getResourceBundle(
+						rbFile   = arguments.rbFile,
+						rbLocale = arguments.rbLocale
+					);
 					// logging
-					if( log.canDebug() ){
-						log.debug( "Loaded bundle: #arguments.rbFile#:#arguments.rbAlias# for locale: #arguments.rbLocale#, forced: #arguments.force#" );
+					if ( log.canDebug() ) {
```
Comment (lmajano) [inline models/ResourceService.cfc:116]: I think this is important to denote in your logs, Can you change this to `info` level please

File: models/ResourceService.cfc [hunk 3]
```diff
@@ -1,439 +1,517 @@
...
+
+			// Loop through property keys and store the values into bundle
+			while ( keys.hasMoreElements() ) {
+				thisKEY                   = keys.nextElement();
+				resourceBundle[ thisKEY ] = rb.handleGetObject( thisKEY );
 			}
-			catch(Any e){
-				fis.close();
-				$rethrow( e );
+		} catch ( Any e ) {
```
Comment (lmajano) [inline models/ResourceService.cfc:267]: Add a `final` block to close the file here instead.  We have this now.

```
finally {
   fis.close();
}

```

File: models/ResourceService.cfc [hunk 4]
```diff
@@ -1,439 +1,517 @@
...
+		if ( !len( rbFullPath ) ) {
+			throw(
+				"The resource bundle file: #rbFilePath# does not exist. Please check your path",
+				"FullPath: #rbFullPath#",
+				"ResourceBundle.InvalidBundlePath"
+			);
+		}
+
+		// read file
+		var fis = createObject( "java", "java.io.FileInputStream" ).init( rbFullPath );
```
Comment (lmajano) [inline models/ResourceService.cfc:316]: This code is the same as in the method above, I think we need to create a private method to encapsulate the loading of the resource file. so both methods can reuse it.

File: models/ResourceService.cfc [hunk 5]
```diff
@@ -1,439 +1,517 @@
...
```
Comment (lmajano) [inline models/ResourceService.cfc:386]: Same as above, thir time we use the same approach

File: models/i18n.cfc [hunk 6]
```diff
@@ -1,970 +1,1094 @@
...
 
-			// are we loading multiple resource bundles? If so, load up their default locale
-			var resourceBundles = variables.controller.getSetting( name="resourceBundles", defaultValue=structNew() );
-			if( structCount( resourceBundles ) ){
-				for( var thisBundle in resourceBundles ){
-					variables.resourceService.loadBundle( rbFile=resourceBundles[ thisBundle ], rbLocale=instance.defaultLocale, rbAlias=lcase( thisBundle ) );
-				}
+		// are we loading multiple resource bundles? If so, load up their default locale
+		var resourceBundles = variables.controller.getSetting( name = "resourceBundles", defaultValue = {} );
+		if ( resourceBundles.count() ) {
```
Comment (lmajano) [inline models/i18n.cfc:60]: you can remove the count(), each won't iterate if empty.  Less code.

File: models/i18n.cfc [hunk 7]
```diff
@@ -1,970 +1,1094 @@
...
+
+	/**
+	 * DateTime format
+	 *
+	 * @thisOffset java epoch offset
+	 * @thisDateFormat FULL=0, LONG=1, MEDIUM=2, SHORT=3
+	 * @thisTimeFormat FULL=0, LONG=1, MEDIUM=2, SHORT=3
+	 * @tz timezone
+	 */
+	string function i18nDateTimeFormat(
```
Comment (lmajano) [inline models/i18n.cfc:312]: We should rename this to just `dateTimeFormat` to make it shorter

File: models/i18n.cfc [hunk 8]
```diff
@@ -1,970 +1,1094 @@
...
+		return tDateFormatter.format( arguments.thisOffset );
+	}
+
+	/**
+	 * Date format
+	 * @thisOffset java epoch offset
+	 * @thisDateFormat FULL=0, LONG=1, MEDIUM=2, SHORT=3
+	 * @tz timezone
+	 */
+	string function i18nDateFormat(
```
Comment (lmajano) [inline models/i18n.cfc:336]: Let's rename to just `dateFormat()`

File: models/i18n.cfc [hunk 9]
```diff
@@ -1,970 +1,1094 @@
...
+	}
+
+	/**
+	 * Time Format
+	 *
+	 * @thisOffset java epoch offset
+	 * @thisTimeFormat FULL=0, LONG=1, MEDIUM=2, SHORT=3
+	 * @tz timezone
+	 */
+	string function i18nTimeFormat(
```
Comment (lmajano) [inline models/i18n.cfc:358]: let's rename to just `timeFormat()`

File: models/i18n.cfc [hunk 10]
```diff
@@ -1,970 +1,1094 @@
...
-				querySetCell(qryLocale,"language",aLocales[i].getDisplayLanguage());
-				querySetCell(qryLocale,"country",aLocales[i].getDisplayCountry());
+	}
+
+	/**
+	 * parses localized date string to datetime object or returns blank if it can't parse
+	 *
+	 * @thisDate
+	 */
+	numeric function i18nDateParse( required string thisDate ) {
```
Comment (lmajano) [inline models/i18n.cfc:453]: we should rename to `dateParse()`

File: models/i18n.cfc [hunk 11]
```diff
@@ -1,970 +1,1094 @@
...
+
+	/**
+	 * i18nDateAdd
+	 *
+	 * @thisOffset
+	 * @thisDatePart
+	 * @dateUnits
+	 * @thisTZ
+	 */
+	numeric function i18nDateAdd(
```
Comment (lmajano) [inline models/i18n.cfc:808]: rename to `dateAdd()`

File: models/i18n.cfc [hunk 12]
```diff
@@ -1,970 +1,1094 @@
...
+
+	/**
+	 * i18nDateDiff
+	 *
+	 * @thisOffset
+	 * @thatOffset
+	 * @thisDatePart
+	 * @thisTZ
+	 */
+	numeric function i18nDateDiff(
```
Comment (lmajano) [inline models/i18n.cfc:864]: rename to `dateDiff`

File: models/ResourceService.cfc (hunk 1 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:39]: removed

File: models/ResourceService.cfc (hunk 2 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:116]: done

File: models/ResourceService.cfc (hunk 3 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:267]: done

File: models/ResourceService.cfc (hunk 4 above)
Comment (wpdebruin) [reply models/ResourceService.cfc:316]: method getResourceFileInputStream created. Used it in 3 different methods

File: models/i18n.cfc (hunk 6 above)
Comment (wpdebruin) [reply models/i18n.cfc:60]: Why didn't I see this :-)
Done

File: models/i18n.cfc (hunk 8 above)
Comment (wpdebruin) [reply models/i18n.cfc:336]: done

File: models/i18n.cfc (hunk 10 above)
Comment (wpdebruin) [reply models/i18n.cfc:453]: Done the same for i18nDateTimeParse

File: models/i18n.cfc [hunk 13]
```diff
@@ -309,7 +307,7 @@ component singleton accessors="true" {
 	 * @thisTimeFormat FULL=0, LONG=1, MEDIUM=2, SHORT=3
 	 * @tz timezone
 	 */
-	string function i18nDateTimeFormat(
+	string function DateTimeFormat(
```
Comment (lmajano) [inline models/i18n.cfc:310]: can you please make sure all method names start with lower case camel case. I think running cfformat will update everything for ya

File: models/ResourceService.cfc [hunk 14]
```diff
@@ -514,4 +461,34 @@ component singleton accessors="true" {
...
+
+	/**
+	 * get Java FileInputStream for resource bundle
+	 * 
+	 * @rbFilePath path + filename for resource, including locale + .properties
+	 * 
+	 * @return java.io.FileInputStream
+	 * @throws ResourceBundle.InvalidBundlePath
+	 */
+	private function getResourceFileInputStream( required string rbFilePath) {
```
Comment (lmajano) [inline models/ResourceService.cfc:476]: beautiful!

**Comment by lmajano** [comment]
Everything looks good. Just run a formatting on it and we are set.

File: models/i18n.cfc (hunk 13 above)
Comment (wpdebruin) [reply models/i18n.cfc]: nope, cfformat doesn't update this. So I did it manually

**Comment by wpdebruin** [comment]
Ok, formatting done, camelCase applied, all set.
So how do I continue now? Are we going to implement changes as described in pull #30 in this request OR should I start a new pull for that?
Pull #30 (the old one) is not useful anymore, because it is against the old code.

**Pull request merged**